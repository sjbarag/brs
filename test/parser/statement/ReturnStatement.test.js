const { Lexeme, BrsTypes, Parser } = require("brs");
const { BrsString } = BrsTypes;
const BrsError = require("../../../lib/Error");

const { EOF } = require("../ParserTests");

describe("parser return statements", () => {
    afterEach(() => BrsError.reset());

    it("parses void returns", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.Function, text: "function", line: 1 },
            { kind: Lexeme.Identifier, text: "foo", line: 1 },
            { kind: Lexeme.LeftParen, text: "(", line: 1 },
            { kind: Lexeme.RightParen, text: ")", line: 1 },
            { kind: Lexeme.Newline, text: "\\n", line: 1 },
            { kind: Lexeme.Return, text: "return", line: 2 },
            { kind: Lexeme.Newline, text: "\\n", line: 2 },
            { kind: Lexeme.EndFunction, text: "end function", line: 3 },
            EOF
        ]);

        expect(BrsError.found()).toBeFalsy();
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed).toMatchSnapshot();
    });

    it("parses literal returns", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.Function, text: "function", line: 1 },
            { kind: Lexeme.Identifier, text: "foo", line: 1 },
            { kind: Lexeme.LeftParen, text: "(", line: 1 },
            { kind: Lexeme.RightParen, text: ")", line: 1 },
            { kind: Lexeme.Newline, text: "\\n", line: 1 },
            { kind: Lexeme.Return, text: "return", line: 2 },
            { kind: Lexeme.String, literal: new BrsString("test"), text: '"test"', line: 2 },
            { kind: Lexeme.Newline, text: "\\n", line: 2 },
            { kind: Lexeme.EndFunction, text: "end function", line: 3 },
            EOF
        ]);

        expect(BrsError.found()).toBeFalsy();
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed).toMatchSnapshot();
    });

    it("parses expression returns", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.Function, text: "function", line: 1 },
            { kind: Lexeme.Identifier, text: "foo", line: 1 },
            { kind: Lexeme.LeftParen, text: "(", line: 1 },
            { kind: Lexeme.RightParen, text: ")", line: 1 },
            { kind: Lexeme.Newline, text: "\\n", line: 1 },
            { kind: Lexeme.Return, text: "return", line: 2 },
            { kind: Lexeme.Identifier, text: "RebootSystem", line: 2 },
            { kind: Lexeme.LeftParen,  text: "(", line: 2 },
            { kind: Lexeme.RightParen, text: ")", line: 2 },
            { kind: Lexeme.Newline, text: "\\n", line: 2 },
            { kind: Lexeme.EndFunction, text: "end function", line: 3 },
            EOF
        ]);

        expect(BrsError.found()).toBeFalsy();
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed).toMatchSnapshot();
    });
});
