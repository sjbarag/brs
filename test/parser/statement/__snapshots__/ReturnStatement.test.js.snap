// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser return statements parses expression returns 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "file": undefined,
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "statements": Array [
          Return {
            "tokens": Object {
              "return": Object {
                "isReserved": true,
                "kind": "Return",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": -9,
                    "line": -9,
                  },
                  "start": Object {
                    "column": -9,
                    "line": -9,
                  },
                },
                "text": "return",
              },
            },
            "type": "Return",
            "value": Call {
              "args": Array [],
              "callee": Variable {
                "name": Object {
                  "isReserved": false,
                  "kind": "Identifier",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": -9,
                      "line": -9,
                    },
                    "start": Object {
                      "column": -9,
                      "line": -9,
                    },
                  },
                  "text": "RebootSystem",
                },
                "type": "Variable",
              },
              "closingParen": Object {
                "isReserved": false,
                "kind": "RightParen",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": -9,
                    "line": -9,
                  },
                  "start": Object {
                    "column": -9,
                    "line": -9,
                  },
                },
                "text": ")",
              },
              "type": "Call",
            },
          },
        ],
        "type": "Block",
      },
      "end": Object {
        "isReserved": false,
        "kind": "EndFunction",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end function",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Function",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "function",
      },
      "parameters": Array [],
      "returns": 10,
      "type": "Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "foo",
    },
    "type": "Function",
  },
]
`;

exports[`parser return statements parses literal returns 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "file": undefined,
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "statements": Array [
          Return {
            "tokens": Object {
              "return": Object {
                "isReserved": true,
                "kind": "Return",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": -9,
                    "line": -9,
                  },
                  "start": Object {
                    "column": -9,
                    "line": -9,
                  },
                },
                "text": "return",
              },
            },
            "type": "Return",
            "value": Literal {
              "_location": undefined,
              "type": "Literal",
              "value": BrsString {
                "kind": 3,
                "value": "test",
              },
            },
          },
        ],
        "type": "Block",
      },
      "end": Object {
        "isReserved": false,
        "kind": "EndFunction",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end function",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Function",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "function",
      },
      "parameters": Array [],
      "returns": 10,
      "type": "Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "foo",
    },
    "type": "Function",
  },
]
`;

exports[`parser return statements parses void returns 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "file": undefined,
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "statements": Array [
          Return {
            "tokens": Object {
              "return": Object {
                "isReserved": true,
                "kind": "Return",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": -9,
                    "line": -9,
                  },
                  "start": Object {
                    "column": -9,
                    "line": -9,
                  },
                },
                "text": "return",
              },
            },
            "type": "Return",
            "value": undefined,
          },
        ],
        "type": "Block",
      },
      "end": Object {
        "isReserved": false,
        "kind": "EndFunction",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end function",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Function",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "function",
      },
      "parameters": Array [],
      "returns": 10,
      "type": "Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "foo",
    },
    "type": "Function",
  },
]
`;
